prow/pjutil/pjutil.go

```diff
@@ -51,14 +53,33 @@ func NewProwJob(spec prowapi.ProwJobSpec, extraLabels, extraAnnotations map[stri
 			Labels:      labels,
 			Annotations: annotations,
 		},
-		Spec: *spec.DeepCopy(),
+		Spec: *specCopy,
 		Status: prowapi.ProwJobStatus{
 			StartTime: metav1.Now(),
 			State:     prowapi.TriggeredState,
 		},
 	}
 }
 
+// setReportDefault sets Slack to false when states to report is an empty slice.
+//
+// `omitempty` is required for fields that are optional, otherwise strict prowjob CRD
+// validation will reject prowjob without the field.
+// For `ReporterConfig.Slack.JobStatesToReport`, it's certainly not a required field
+// for all prowjobs.
+// However, when omitempty is presented, `JobStatesToReport: []` roundtrip into
+// `JobStatesToReport: nil`, which prevents a prowjob from overriding global Slack
+// report configuration(ref: https://github.com/kubernetes/test-infra/issues/22888#issuecomment-881513368).
+// Use a boolean instead so that we can use `omitempty`.
+func setReportDefault(spec *prowapi.ProwJobSpec) {
```

Comment:
This doesn't do anything, because a non-pointer bool field with `omitempty` will be omitted in the serializiation when the value is `false` as that is the default for an unititialized boolean.

I think we need to make it a pointer field to differentiate `uninitialized` from `false` and also have an `else` block in the second condition here that sets it to `true`?

Reply:
good point, updated the PR, as well as some unit tests